File: tools/project/cabal_test.pl
#!/usr/bin/env perl

use strict;
use warnings FATAL => 'all';
use utf8;

sub slurp {
   local $/;
   open my $fh, '<', $_[0] or die "'$_[0]': $!";
   <$fh>
}

my ($build_file, $cabal_file) = @ARGV;
my $build = slurp $build_file;
my $cabal = slurp $cabal_file;

# Check package name consistency.
my ($cabal_name) = $cabal =~ m/^name:\s+([a-zA-Z0-9_-]+)\n/
   or die "name field must be defined in '$cabal_file'\n";
my ($build_name) = $build =~ m/haskell_library\(\n    name = "hs-([a-zA-Z0-9_-]+)",\n/
   or die "a haskell_library starting with 'hs-' must be defined in '$build_file'\n";

die "$cabal_file ($cabal_name) and $build_file ($build_name) disagree on package name\n"
   if $cabal_name ne $build_name;

# Check test target name.
my ($test_name) = $build =~ m/hspec_test\(\n    name = "([a-zA-Z0-9_-]+)",\n/
   or die "a hspec_test must be defined in '$build_file'\n";

die "hspec_test should be named 'testsuite' but is '$test_name'\n"
   if $test_name ne "testsuite";

# Check version consistency.
my ($cabal_version) = $cabal =~ m/\nversion:\s+([0-9.]+)\n/
   or die "version field must be defined in '$cabal_file'\n";
my ($build_version) = $build =~ m/\n    version = "([0-9.]+)",\n/;
($build_version) = $build =~ m/\nVERSION = "([0-9.]+)"\n/
   unless $build_version;
$build_version or die "version field must be defined in '$build_file'\n";

die "$cabal_file ($cabal_version) and $build_file ($build_version) disagree on package version\n"
   if $cabal_version ne $build_version;

# Build dependencies consistency.
sub compare_deps {
   my ($cabal_decl, $build_decl) = @_;

   my ($cabal_library_depends) = $cabal =~ m/\n$cabal_decl\n(?:.|\n)*?  build-depends:\s*([^:]+)\n(?:\n|$)/
      or die "$cabal_decl.build-depends field must be defined in '$cabal_file'\n";
   my @cabal_library_depends = split /\s*,\s*/, $cabal_library_depends;
   s/ .*// for @cabal_library_depends;
   @cabal_library_depends = sort @cabal_library_depends;

#   print "$cabal_decl: $_\n" for @cabal_library_depends;

   my ($build_library_depends) = $build =~ m/\n$build_decl\(\n(?:.|\n)*?    deps = \[\n*\s*([^\]]+)\]/
      or die "$build_decl.deps must be defined in '$build_file'\n";
   my @build_library_depends = grep { not m/"\/\/c-/ } split /\s*,\s*/, $build_library_depends;
   s/"(?::|\/\/)hs-(.*)"/$1/g for @build_library_depends;
   s/hazel_library\("(.*)"\)/$1/g for @build_library_depends;
   @build_library_depends = sort @build_library_depends;

#   print "$build_decl $_\n" for @build_library_depends;

   my %in_both;
   for my $build_dep (@build_library_depends) {
      $in_both{$build_dep} = 1 if grep { $_ eq $build_dep } @cabal_library_depends;
   }
#   print "both $_\n" for sort keys %in_both;

   @cabal_library_depends = grep { not $in_both{$_} } @cabal_library_depends;
   @build_library_depends = grep { not $in_both{$_} } @build_library_depends;
   die "$cabal_decl.build-depends in $cabal_file not present in $build_decl.deps in $build_file: @cabal_library_depends\n"
      if @cabal_library_depends;
   die "$build_decl.deps in $build_file not present in $cabal_decl.build-depends in $cabal_file: @build_library_depends\n"
      if @build_library_depends;
}

compare_deps "library", "haskell_library";
compare_deps "test-suite testsuite", "hspec_test";

File: tools/project/cabal_test.py
#!/usr/bin/env python3

import collections
import json
import re
import sys


def parse_bazel(build):
    data = {}

    def store(rule):
        def f(name, **kwargs):
            data[name] = kwargs
            data[name]["rule"] = rule
        return f

    # pylint: disable=exec-used
    exec(open(build).read(), {
        "load": lambda bzl, *symbols: None,
        "project": lambda **kwargs: None,
        "alex_lexer": lambda **kwargs: None,
        "happy_parser": lambda **kwargs: None,
        "glob": lambda include, exclude=[]: [],
        "hazel_library": lambda name: name,
        "haskell_library": store("haskell_library"),
        "hspec_test": store("hspec_test"),
    })
    return json.loads(json.dumps(data))


def parse_cabal(cabal):
    def rec_dd(): return collections.defaultdict(rec_dd)
    data = rec_dd()
    section = data['']['']
    key = None
    value = None

    with open(cabal, "r") as fh:
        for line in fh.readlines():
            line = line.rstrip()
            res = re.match(r"\s*([a-z-]+):\s*(.*)$", line)
            if res:
                if key:
                    values = value.split(", ")
                    section[key] = values if len(values) > 1 else value
                    key = None
                key, value = res.group(1, 2)
                continue
            res = re.match(r"([a-z-]+)\s*([a-z-]*)$", line)
            if res:
                if key:
                    values = value.split(", ")
                    section[key] = values if len(values) > 1 else value
                    key = None
                section = data[res.group(1)][res.group(2)]
                section["__name__"] = line
                continue
            value += line.strip()
        if key:
            values = value.split(", ")
            section[key] = values if len(values) > 1 else value
            key = None
    return json.loads(json.dumps(data))


def resolve_bazel_deps(bazel, data, name):
    def go(name):
        deps = []
        rule = data.get(name, None)
        if not rule:
            print("missing rule for '{name}' in {bazel}".format(
                name=name, bazel=bazel))
            sys.exit(1)
        for dep in rule["deps"]:
            if dep.startswith(":hs-"):
                deps.append(dep[4:])
            elif dep.startswith("//hs-"):
                deps.append(dep[5:])
            elif dep.startswith("//c-"):
                pass  # ignore native deps for now
            elif dep.startswith(":"):
                if name != "testsuite":
                    deps.extend(go(dep[1:]))
            else:
                deps.append(dep)
        return deps
    return set(go(name))


def must_list(maybe_list):
    if isinstance(maybe_list, list):
        return maybe_list
    else:
        return [maybe_list]


def check_deps(bazel, cabal, bazel_data, bazel_name, cabal_section):
    bazel_deps = resolve_bazel_deps(bazel, bazel_data, bazel_name)
    cabal_deps = {dep.split(" ")[0]
                  for dep in must_list(cabal_section["build-depends"])}

    cabal_missing = sorted(bazel_deps - cabal_deps)
    section = cabal_section['__name__']
    if cabal_missing:
        print("{cabal} is missing deps for section "
              "'{section}' from {bazel} '{bazel_name}': "
              "{cabal_missing}".format(
                  cabal=cabal,
                  section=section,
                  bazel=bazel,
                  bazel_name=bazel_name,
                  cabal_missing=cabal_missing,
              ))
        sys.exit(1)

    bazel_missing = sorted(cabal_deps - bazel_deps)
    if bazel_missing:
        print("{bazel} is missing deps for '{bazel_name}' from "
              "{cabal} section '{section}: "
              "{bazel_missing}".format(
                  bazel=bazel,
                  bazel_name=bazel_name,
                  cabal=cabal,
                  section=section,
                  bazel_missing=bazel_missing,
              ))
        sys.exit(1)


def check_version(root, bazel_data, package_name):
    cabal_version = root["version"]
    bazel_version = bazel_data["hs-" + package_name]["version"]
    if cabal_version != bazel_version:
        print("bazel version ({bazel_version}) and cabal version "
              "({cabal_version}) disagree".format(
                  bazel_version=bazel_version,
                  cabal_version=cabal_version,
              ))
        sys.exit(1)


def main(args):
    bazel, cabal = args

    bazel_data = parse_bazel(bazel)
    cabal_data = parse_cabal(cabal)

    root = cabal_data[""][""]
    package_name = root["name"]

    check_deps(bazel, cabal, bazel_data,
               "hs-" + package_name, cabal_data["library"][""])
    check_deps(bazel, cabal, bazel_data,
               "testsuite", cabal_data["test-suite"]["testsuite"])

    check_version(root, bazel_data, package_name)


main(sys.argv[1:])
